zk.js/src/transaction/solanaTransaction.ts

```diff
@@ -371,22 +408,124 @@ export async function sendAndConfirmShieldedTransaction({
     solanaTransactionInputs.systemProof.parsedPublicInputsObject,
     remainingMerkleTreeAccounts,
   );
+  const accounts = prepareAccounts({
+    transactionAccounts: publicTransactionVariables.accounts,
+    eventMerkleTreePubkey: eventMerkleTree,
+    relayerRecipientSol,
+    signer: publicTransactionVariables.accounts.relayerPublicKey,
+  });
 
   // createSolanaInstructionsWithAccounts
-  const instructions = await createSolanaInstructions(
+  const instructions = await createSolanaInstructions({
+    action,
     rootIndex,
-    solanaTransactionInputs.systemProof,
+    systemProof,
     remainingSolanaAccounts,
-    solanaTransactionInputs.transaction,
-    solanaTransactionInputs.pspTransactionInput
-      ? solanaTransactionInputs.pspTransactionInput.verifierIdl
-      : solanaTransactionInputs.transaction.verifierIdl,
-    solanaTransactionInputs.pspTransactionInput,
-    solanaTransactionInputs.pspProof,
-  );
+    accounts,
+    systemPspIdl,
+    pspIdl: pspTransactionInput?.verifierIdl,
+    pspTransactionInput,
+    pspProof,
+    publicTransactionVariables,
+  });
 
   const txHash = await provider.sendAndConfirmShieldedTransaction(instructions);
   const relayerMerkleTreeUpdateResponse = "notPinged";
 
   return { txHash, response: relayerMerkleTreeUpdateResponse };
 }
+
+// TODO: unify event Merkle tree and transaction Merkle tree so that only one is passed
+export function prepareAccounts({
+  transactionAccounts,
+  eventMerkleTreePubkey,
+  signer,
+  relayerRecipientSol,
+  verifierState,
+}: {
+  transactionAccounts: TransactionAccounts;
+  eventMerkleTreePubkey: PublicKey;
+  signer?: PublicKey;
+  relayerRecipientSol?: PublicKey;
+  verifierState?: PublicKey;
+}): lightAccounts {
+  const {
+    senderSol,
+    senderSpl,
+    recipientSol,
+    recipientSpl,
+    relayerPublicKey,
+    pspId,
+    systemPspId,
+  } = transactionAccounts;
+  const verifierProgramId = pspId ? pspId : systemPspId;
+  if (!signer) {
+    signer = relayerPublicKey;
+  }
+  if (!verifierState) {
+    verifierState = PublicKey.findProgramAddressSync(
+      [signer.toBytes(), utils.bytes.utf8.encode("VERIFIER_STATE")],
+      verifierProgramId,
+    )[0];
```

Comment:
does it make sense to move it to a helper function e.g. getNullifierPda ?

Reply:
in fact we already have one lol